# E-commerce Operational Performance Analysis

## Data Understanding & Data Preparation

This project analyzes e-commerce operational data to evaluate order performance, fulfillment efficiency, cancellations, returns, operational bottlenecks, and data consistency.

### Objectives
- Understand the structure and grain of the operational dataset
- Assess and improve data quality
- Monitor order completion and cancellation performance
- Measure order processing efficiency
- Identify operational bottlenecks and anomalies
- Perform operational reconciliation checks
- Prepare clean datasets for dashboard development

In [341]:
import pandas as pd
import numpy as np
import os

In [342]:
file_path = "/content/VidronaOct2025.xlsx"

df_order = pd.read_excel( file_path, sheet_name='orders')

In [343]:
df_order.shape

(3266, 50)

### Initial Observation

The raw order dataset contains 3,266 rows and 50 columns.

At this stage, the number of rows should not be interpreted as the number of unique orders because the dataset grain has not yet been determined.

In [344]:
df_order.columns

Index(['No. Pesanan', 'Status Pesanan', 'Alasan Pembatalan',
       'Status Pembatalan/ Pengembalian', 'No. Resi', 'Opsi Pengiriman',
       'Antar ke counter/ pick-up',
       'Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)',
       'Waktu Pengiriman Diatur', 'Waktu Pesanan Dibuat',
       'Waktu Pembayaran Dilakukan', 'Tipe Pesanan', 'Metode Pembayaran',
       'SKU Induk', 'Nama Produk', 'Nomor Referensi SKU', 'Nama Variasi',
       'Harga Awal', 'Harga Setelah Diskon', 'Jumlah', 'Returned quantity',
       'Subtotal Pesanan', 'Total Diskon', 'Diskon Dari Penjual',
       'Diskon Dari Shopee', 'Berat Produk', 'Jumlah Produk di Pesan',
       'Total Berat', 'Voucher Ditanggung Penjual', 'Cashback Koin',
       'Voucher Ditanggung Shopee', 'Paket Diskon',
       'Paket Diskon (Diskon dari Shopee)',
       'Paket Diskon (Diskon dari Penjual)', 'Potongan Koin Shopee',
       'Diskon Kartu Kredit', 'Ongkos Kirim Dibayar oleh Pembeli',
       'Estimasi Potongan Biaya Pengiri

In [345]:
for i, column in enumerate(df_order.columns, start=1):
  print (i, column)

1 No. Pesanan
2 Status Pesanan
3 Alasan Pembatalan
4 Status Pembatalan/ Pengembalian
5 No. Resi
6 Opsi Pengiriman
7 Antar ke counter/ pick-up
8 Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)
9 Waktu Pengiriman Diatur
10 Waktu Pesanan Dibuat
11 Waktu Pembayaran Dilakukan
12 Tipe Pesanan
13 Metode Pembayaran
14 SKU Induk
15 Nama Produk
16 Nomor Referensi SKU
17 Nama Variasi
18 Harga Awal
19 Harga Setelah Diskon
20 Jumlah
21 Returned quantity
22 Subtotal Pesanan
23 Total Diskon
24 Diskon Dari Penjual
25 Diskon Dari Shopee
26 Berat Produk
27 Jumlah Produk di Pesan
28 Total Berat
29 Voucher Ditanggung Penjual
30 Cashback Koin
31 Voucher Ditanggung Shopee
32 Paket Diskon
33 Paket Diskon (Diskon dari Shopee)
34 Paket Diskon (Diskon dari Penjual)
35 Potongan Koin Shopee
36 Diskon Kartu Kredit
37 Ongkos Kirim Dibayar oleh Pembeli
38 Estimasi Potongan Biaya Pengiriman
39 Ongkos Kirim Pengembalian Barang
40 Total Pembayaran
41 Perkiraan Ongkos Kirim
42 Catatan dari Pembeli
43 Catata

### Data types

In [346]:
df_order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3266 entries, 0 to 3265
Data columns (total 50 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   No. Pesanan                                                   3266 non-null   object 
 1   Status Pesanan                                                3266 non-null   object 
 2   Alasan Pembatalan                                             511 non-null    object 
 3   Status Pembatalan/ Pengembalian                               36 non-null     object 
 4   No. Resi                                                      2824 non-null   object 
 5   Opsi Pengiriman                                               3266 non-null   object 
 6   Antar ke counter/ pick-up                                     2824 non-null   object 
 7   Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)  3143 no

In [347]:
df_order.isna().sum()

,0
No. Pesanan,0
Status Pesanan,0
Alasan Pembatalan,2755
Status Pembatalan/ Pengembalian,3230
No. Resi,442
Opsi Pengiriman,0
Antar ke counter/ pick-up,442
Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan),123
Waktu Pengiriman Diatur,448
Waktu Pesanan Dibuat,0


### Missing values

In [348]:
missing_values = pd.DataFrame({
    'missing_count' : df_order.isna().sum(),
    'missing_percentage' : (df_order.isna().sum() / len(df_order) * 100).round(2)}
)

In [349]:
missing_values

,missing_count,missing_percentage
No. Pesanan,0,0.00
Status Pesanan,0,0.00
Alasan Pembatalan,2755,84.35
Status Pembatalan/ Pengembalian,3230,98.90
No. Resi,442,13.53
Opsi Pengiriman,0,0.00
Antar ke counter/ pick-up,442,13.53
Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan),123,3.77
Waktu Pengiriman Diatur,448,13.72
Waktu Pesanan Dibuat,0,0.00


In [350]:
missing_values.sort_values('missing_percentage', ascending=False)

,missing_count,missing_percentage
Tipe Pesanan,3266,100.00
Catatan,3266,100.00
Status Pembatalan/ Pengembalian,3230,98.90
Catatan dari Pembeli,3205,98.13
SKU Induk,3049,93.36
Alasan Pembatalan,2755,84.35
Waktu Pesanan Selesai,511,15.65
Waktu Pengiriman Diatur,448,13.72
Antar ke counter/ pick-up,442,13.53
No. Resi,442,13.53


### Exact duplicate

In [351]:
print(
    'Exact Duplicated Rows:', df_order.duplicated().sum()
)

Exact Duplicated Rows: 0


### Looking for dataset grain

### Unique Order ID

In [352]:
print("Total rows:", len(df_order))

Total rows: 3266


In [353]:
print("Unique order IDs:", df_order['No. Pesanan'].nunique())

Unique order IDs: 3153


In [354]:
order_counts = (df_order['No. Pesanan'].value_counts())

In [355]:
order_counts_2 = (df_order['Nama Produk'].value_counts())

In [356]:
order_counts_2.head(10)

,count
Nama Produk,
"Tongsis Tripod HP 3 in 1 dengan Lampu LED - VIDRONA ELK-05, Multifungsi & Portable",1331
"VIDRONA ELK-50 Belt Holder HP Tali Dada 360° untuk Motovlog, Video Motor, dan Mount Kamera GoPro",322
Tripod Ring Light LED 26CM - Stand HP 1 Meter Fleksibel Lampu untuk Live & Makeup,210
VIDRONA ELK-27 Paket Komplit Ring Light LED 26CM + Tripod 1 Meter,163
VIDRONA Sarung Tripod Tas Tripod 2 Meter Tas Light Stand Kamera Bahan Tebal Premium Tahan Lama,157
VIDRONA AKS-18 Stand Holder HP Meja Fleksibel 360°,107
VIDRONA AKS-156 Stand X Banner 60x160 cm Rangka Promosi Display Lipat untuk Indoor Outdoor,74
VIDRONA ELK-05 Tongsis Tripod dengan Lampu LED untuk Selfie dan Video 360º,63
X Banner Stand Tripod Banner 60x160cm Rangka Spanduk Standing Poster,61


In [357]:
len(order_counts_2)

78

In [358]:
order_counts.head(10)

,count
No. Pesanan,
2510156W2XQ23K,4
25100463D8080U,3
2510131FYHB1UR,3
251009M4N8G4SA,3
2510240CDAJSCX,3
2510046WDYJ732,3
251008H0U4EXS6,3
2510034K9R2R0W,3
251018CWX6VXGC,3


In [359]:
len(order_counts)

3153

In [360]:
order_counts[order_counts > 1].head(10)

,count
No. Pesanan,
2510156W2XQ23K,4
25100463D8080U,3
2510131FYHB1UR,3
251009M4N8G4SA,3
2510240CDAJSCX,3
2510046WDYJ732,3
251008H0U4EXS6,3
2510034K9R2R0W,3
251018CWX6VXGC,3


In [361]:
len(order_counts[order_counts > 1])

103

### Investigation of a repeat order

In [362]:
duplicate_orders = order_counts[order_counts > 1]

In [363]:
duplicate_orders.index[0]

'2510156W2XQ23K'

In [364]:
df_order.columns

Index(['No. Pesanan', 'Status Pesanan', 'Alasan Pembatalan',
       'Status Pembatalan/ Pengembalian', 'No. Resi', 'Opsi Pengiriman',
       'Antar ke counter/ pick-up',
       'Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)',
       'Waktu Pengiriman Diatur', 'Waktu Pesanan Dibuat',
       'Waktu Pembayaran Dilakukan', 'Tipe Pesanan', 'Metode Pembayaran',
       'SKU Induk', 'Nama Produk', 'Nomor Referensi SKU', 'Nama Variasi',
       'Harga Awal', 'Harga Setelah Diskon', 'Jumlah', 'Returned quantity',
       'Subtotal Pesanan', 'Total Diskon', 'Diskon Dari Penjual',
       'Diskon Dari Shopee', 'Berat Produk', 'Jumlah Produk di Pesan',
       'Total Berat', 'Voucher Ditanggung Penjual', 'Cashback Koin',
       'Voucher Ditanggung Shopee', 'Paket Diskon',
       'Paket Diskon (Diskon dari Shopee)',
       'Paket Diskon (Diskon dari Penjual)', 'Potongan Koin Shopee',
       'Diskon Kartu Kredit', 'Ongkos Kirim Dibayar oleh Pembeli',
       'Estimasi Potongan Biaya Pengiri

In [365]:
if len(duplicate_orders) > 0:
  duplicate_order_id = duplicate_orders.index[0]

  display(df_order[df_order['No. Pesanan'] == duplicate_order_id]
   [[
       'No. Pesanan',
       'No. Resi',
       'Nama Produk',
       'Nama Variasi']]
          )

,No. Pesanan,No. Resi,Nama Produk,Nama Variasi
1668,2510156W2XQ23K,SPXID05019818503A,VIDRONA Stand Holder HP Lipat Universal Duduka...,Phone Holder Foldabl
1669,2510156W2XQ23K,SPXID05019818503A,VIDRONA ELK-11 Cute Mini Fan Kipas Lipat Porta...,Blue
1670,2510156W2XQ23K,SPXID05019818503A,VIDRONA ELK-153 Tripod Mini Portable 4 Level T...,Black
1671,2510156W2XQ23K,SPXID05019818503A,VIDRONA Satu Set Ringlight 33cm Tripod 2 Meter...,Tripod 2M+R33cm


### Dataset Grain Conclusion

The dataset is stored at the **order-item level**, meaning that one row represents one product/item within an order rather than one unique order.

The dataset contains 3,266 rows but only 3,153 unique order IDs. A total of 103 orders appear across multiple rows.

Investigation of repeated order IDs confirms that these rows are not duplicate records. Instead, they represent different products or product variations purchased within the same order.

Therefore:

- `No. Pesanan` is not unique at the row level.
- One order may contain multiple products or variations.
- Order-level KPIs must be calculated using unique order IDs to prevent double counting.
- Product and item-level metrics can be calculated using individual rows.

### Distribusi jumlah item-row per order

In [366]:
order_counts

,count
No. Pesanan,
2510156W2XQ23K,4
25100463D8080U,3
2510131FYHB1UR,3
251009M4N8G4SA,3
2510240CDAJSCX,3
...,...
251011RM5SMXN6,1
251011RNJMXFMY,1
251011RNYQK28P,1


In [367]:
multi_item_orders = order_counts[order_counts > 1]

In [368]:
print( 'Orders with multiple rows:', len(multi_item_orders))

Orders with multiple rows: 103


In [369]:
print("Maximum rows in one order:", multi_item_orders.max())

Maximum rows in one order: 4


This observation also validate the data grain

## Data Quality Assessment

This section evaluates the quality and consistency of the raw operational dataset before cleaning and transformation.

The assessment focuses on:

- Missing values
- Duplicate records
- Order status consistency
- Cancellation and return information
- Fulfillment information
- Data types
- Logical relationships between operational fields

### Checking Status Order

In [370]:
df_order['Status Pesanan'].value_counts(dropna=False)

,count
Status Pesanan,
Selesai,2755
Batal,511


In [371]:
df_order['Status Pesanan'].value_counts(normalize=True, dropna=False).mul(100).round(2)

,proportion
Status Pesanan,
Selesai,84.35
Batal,15.65


### Cancellation & Return Status

In [372]:
df_order['Status Pembatalan/ Pengembalian'].value_counts(dropna=False)

,count
Status Pembatalan/ Pengembalian,
NaN,3230
Permintaan Disetujui,35
Masalah Diselesaikan,1


In [373]:
df_order['Alasan Pembatalan'].value_counts(dropna=False)

,count
Alasan Pembatalan,
NaN,2755
Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,124
Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada,112
Dibatalkan oleh Pembeli. Alasan: Need to change delivery address,85
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar,77
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal,61
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman,22
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan,10
Dibatalkan oleh Pembeli. Alasan: Penjual tidak responsif terhadap pertanyaan Pembeli,5


### Investigasi missing fulfillment

In [374]:
df_order.columns

Index(['No. Pesanan', 'Status Pesanan', 'Alasan Pembatalan',
       'Status Pembatalan/ Pengembalian', 'No. Resi', 'Opsi Pengiriman',
       'Antar ke counter/ pick-up',
       'Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)',
       'Waktu Pengiriman Diatur', 'Waktu Pesanan Dibuat',
       'Waktu Pembayaran Dilakukan', 'Tipe Pesanan', 'Metode Pembayaran',
       'SKU Induk', 'Nama Produk', 'Nomor Referensi SKU', 'Nama Variasi',
       'Harga Awal', 'Harga Setelah Diskon', 'Jumlah', 'Returned quantity',
       'Subtotal Pesanan', 'Total Diskon', 'Diskon Dari Penjual',
       'Diskon Dari Shopee', 'Berat Produk', 'Jumlah Produk di Pesan',
       'Total Berat', 'Voucher Ditanggung Penjual', 'Cashback Koin',
       'Voucher Ditanggung Shopee', 'Paket Diskon',
       'Paket Diskon (Diskon dari Shopee)',
       'Paket Diskon (Diskon dari Penjual)', 'Potongan Koin Shopee',
       'Diskon Kartu Kredit', 'Ongkos Kirim Dibayar oleh Pembeli',
       'Estimasi Potongan Biaya Pengiri

Missing tracking number

In [375]:
df_order[df_order['No. Resi'].isna()]['Status Pesanan'].value_counts(dropna=False)

,count
Status Pesanan,
Batal,442


Missing shipping arranged time

In [376]:
df_order[df_order['Waktu Pengiriman Diatur'].isna()]['Status Pesanan'].value_counts(dropna=False)

,count
Status Pesanan,
Batal,448


Missing completion time

In [377]:
df_order[df_order['Waktu Pesanan Selesai'].isna()]['Status Pesanan'].value_counts(dropna=False)

,count
Status Pesanan,
Batal,511


Missing fulfillment information is coming from cancelled/uncompleted orders

### Payment timestamp investigation

In [378]:
df_order['Waktu Pembayaran Dilakukan'].isna().sum()

np.int64(0)

In [379]:
df_order[
    df_order["Waktu Pesanan Selesai"].isna()
]['Status Pesanan'].value_counts(dropna=False)

,count
Status Pesanan,
Batal,511


## Check Return Quantity

In [380]:
df_order['Returned quantity'].nunique()

2

In [381]:
df_order[df_order['Returned quantity'] == 1]['Nama Produk'].value_counts(dropna=False)

,count
Nama Produk,
"Tongsis Tripod HP 3 in 1 dengan Lampu LED - VIDRONA ELK-05, Multifungsi & Portable",17
VIDRONA ELK-27 Paket Komplit Ring Light LED 26CM + Tripod 1 Meter,4
VIDRONA ELK-99 Stand Mic Meja Boom Arm Microphone untuk Podcast Recording,2
Tripod Ring Light LED 26CM - Stand HP 1 Meter Fleksibel Lampu untuk Live & Makeup,2
VIDRONA AKS-18 Stand Holder HP Meja Fleksibel 360°,1
VIDRONA ELK-46 Tripod 140CM + Ring Light LED 26CM Stand HP Untuk Tiktok Live Streaming,1
X Banner Stand Tripod Banner 60x160cm Rangka Spanduk Standing Poster,1
VIDRONA AKS-18 Stand Holder HP Phone Holder Stand Putar 360°,1
VIDRONA Sarung Tripod Tas Tripod 2 Meter Tas Light Stand Kamera Bahan Tebal Premium Tahan Lama,1


In [382]:
df_order['Returned quantity'].value_counts(dropna=False)

,count
Returned quantity,
0,3230
1,36


In [383]:
df_order.columns

Index(['No. Pesanan', 'Status Pesanan', 'Alasan Pembatalan',
       'Status Pembatalan/ Pengembalian', 'No. Resi', 'Opsi Pengiriman',
       'Antar ke counter/ pick-up',
       'Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)',
       'Waktu Pengiriman Diatur', 'Waktu Pesanan Dibuat',
       'Waktu Pembayaran Dilakukan', 'Tipe Pesanan', 'Metode Pembayaran',
       'SKU Induk', 'Nama Produk', 'Nomor Referensi SKU', 'Nama Variasi',
       'Harga Awal', 'Harga Setelah Diskon', 'Jumlah', 'Returned quantity',
       'Subtotal Pesanan', 'Total Diskon', 'Diskon Dari Penjual',
       'Diskon Dari Shopee', 'Berat Produk', 'Jumlah Produk di Pesan',
       'Total Berat', 'Voucher Ditanggung Penjual', 'Cashback Koin',
       'Voucher Ditanggung Shopee', 'Paket Diskon',
       'Paket Diskon (Diskon dari Shopee)',
       'Paket Diskon (Diskon dari Penjual)', 'Potongan Koin Shopee',
       'Diskon Kartu Kredit', 'Ongkos Kirim Dibayar oleh Pembeli',
       'Estimasi Potongan Biaya Pengiri

In [384]:
df_order[df_order["Returned quantity"] > 0
][[
        "No. Pesanan",
        "Status Pesanan",
        "Status Pembatalan/ Pengembalian",
        "Nama Produk",
        "Nama Variasi",
        "Jumlah",
        "Returned quantity"
    ]
].head(20)

,No. Pesanan,Status Pesanan,Status Pembatalan/ Pengembalian,Nama Produk,Nama Variasi,Jumlah,Returned quantity
57,2510010309H0F4,Selesai,Permintaan Disetujui,VIDRONA ELK-99 Stand Mic Meja Boom Arm Microph...,Mic Stand,1,1
82,2510010H1H8Q2A,Selesai,Permintaan Disetujui,VIDRONA AKS-18 Stand Holder HP Meja Fleksibel ...,NaN,1,1
209,2510034SXHABBC,Selesai,Permintaan Disetujui,VIDRONA ELK-46 Tripod 140CM + Ring Light LED 2...,ALL IN,1,1
412,2510059XU8W6HV,Selesai,Permintaan Disetujui,VIDRONA ELK-27 Paket Komplit Ring Light LED 26...,Tripod 1M+Ring Light,1,1
413,2510059XW6VUQ6,Selesai,Masalah Diselesaikan,Tripod Ring Light LED 26CM - Stand HP 1 Meter ...,Tripod 1M+Ring Light,1,1
781,251008HDT6BBQB,Selesai,Permintaan Disetujui,VIDRONA ELK-27 Paket Komplit Ring Light LED 26...,Tripod 1M+Ring Light,1,1
1096,251011R53Q1NGK,Selesai,Permintaan Disetujui,VIDRONA ELK-27 Paket Komplit Ring Light LED 26...,Tripod 1M+Ring Light,1,1
1135,251011RYPDPB52,Selesai,Permintaan Disetujui,Tongsis Tripod HP 3 in 1 dengan Lampu LED - VI...,Black + LED,1,1
1152,251011SBVBJ808,Selesai,Permintaan Disetujui,Tongsis Tripod HP 3 in 1 dengan Lampu LED - VI...,Black + LED,1,1
1155,251011SD5S6C3Q,Selesai,Permintaan Disetujui,VIDRONA ELK-27 Paket Komplit Ring Light LED 26...,Tripod 1M+Ring Light,1,1


### Data Type Assessment

In [385]:
datetime_columns = [
    "Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan)",
    "Waktu Pengiriman Diatur",
    "Waktu Pesanan Dibuat",
    "Waktu Pembayaran Dilakukan",
    "Waktu Pesanan Selesai"
]

df_order[datetime_columns].dtypes

,0
Pesanan Harus Dikirimkan Sebelum (Menghindari keterlambatan),object
Waktu Pengiriman Diatur,object
Waktu Pesanan Dibuat,object
Waktu Pembayaran Dilakukan,object
Waktu Pesanan Selesai,object


### Numerical sanity check

In [386]:
numeric_columns = (df_order.select_dtypes(include=np.number).columns.tolist()
)

numeric_columns

['Tipe Pesanan',
 'Harga Awal',
 'Harga Setelah Diskon',
 'Jumlah',
 'Returned quantity',
 'Diskon Dari Shopee',
 'Jumlah Produk di Pesan',
 'Voucher Ditanggung Penjual',
 'Cashback Koin',
 'Voucher Ditanggung Shopee',
 'Paket Diskon (Diskon dari Shopee)',
 'Paket Diskon (Diskon dari Penjual)',
 'Potongan Koin Shopee',
 'Diskon Kartu Kredit',
 'Ongkos Kirim Dibayar oleh Pembeli',
 'Estimasi Potongan Biaya Pengiriman',
 'Ongkos Kirim Pengembalian Barang',
 'Perkiraan Ongkos Kirim',
 'Catatan']

In [387]:
df_order[
    numeric_columns
].describe().T

,count,mean,std,min,25%,50%,75%,max
Tipe Pesanan,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Harga Awal,3266.0,180.601676,96.018922,3.00,150.0,170.00,170.00,950.0
Harga Setelah Diskon,3266.0,69.777761,43.201619,2.97,48.4,58.90,71.30,470.0
Jumlah,3266.0,1.075628,0.644049,1.00,1.0,1.00,1.00,20.0
Returned quantity,3266.0,0.011023,0.104425,0.00,0.0,0.00,0.00,1.0
Diskon Dari Shopee,3266.0,0.000000,0.000000,0.00,0.0,0.00,0.00,0.0
Jumlah Produk di Pesan,3266.0,1.154930,0.733701,1.00,1.0,1.00,1.00,20.0
Voucher Ditanggung Penjual,3266.0,73.429271,176.797932,0.00,0.0,0.00,0.00,500.0
Cashback Koin,3266.0,0.000000,0.000000,0.00,0.0,0.00,0.00,0.0
Voucher Ditanggung Shopee,3266.0,9.540665,53.796714,0.00,0.0,2.66,8.89,999.0


this  to detect this condition
- quantity < 0
- returned quantity < 0
- returned quantity > quantity
- price < 0

## Data Cleaning and Standardization

Based on the data quality assessment, no exact duplicate records were identified.

Missing operational values will not be removed automatically because some missing values are structurally related to cancelled orders.

The cleaning process focuses on:

- Preserving the original raw dataset
- Standardizing column names
- Simplifying frequently used operational fields
- Converting date and time fields
- Validating data type conversions
- Preserving legitimate missing values

### Make a working copy

In [388]:
df_clean = df_order.copy()

In [389]:
print ('Raw Dataset:', df_order.shape)
print ('Working Dataset:', df_clean.shape)

Raw Dataset: (3266, 50)
Working Dataset: (3266, 50)


### Standardization of column names

In [390]:
df_clean.columns = (
    df_clean.columns
    .str.strip()
    .str.lower()
    .str.replace(r"[./()]", "", regex=True)
    .str.replace(r"\s+", "_", regex=True)
)

In [391]:
for i, column in enumerate(df_clean.columns):
  print (i, column)


0 no_pesanan
1 status_pesanan
2 alasan_pembatalan
3 status_pembatalan_pengembalian
4 no_resi
5 opsi_pengiriman
6 antar_ke_counter_pick-up
7 pesanan_harus_dikirimkan_sebelum_menghindari_keterlambatan
8 waktu_pengiriman_diatur
9 waktu_pesanan_dibuat
10 waktu_pembayaran_dilakukan
11 tipe_pesanan
12 metode_pembayaran
13 sku_induk
14 nama_produk
15 nomor_referensi_sku
16 nama_variasi
17 harga_awal
18 harga_setelah_diskon
19 jumlah
20 returned_quantity
21 subtotal_pesanan
22 total_diskon
23 diskon_dari_penjual
24 diskon_dari_shopee
25 berat_produk
26 jumlah_produk_di_pesan
27 total_berat
28 voucher_ditanggung_penjual
29 cashback_koin
30 voucher_ditanggung_shopee
31 paket_diskon
32 paket_diskon_diskon_dari_shopee
33 paket_diskon_diskon_dari_penjual
34 potongan_koin_shopee
35 diskon_kartu_kredit
36 ongkos_kirim_dibayar_oleh_pembeli
37 estimasi_potongan_biaya_pengiriman
38 ongkos_kirim_pengembalian_barang
39 total_pembayaran
40 perkiraan_ongkos_kirim
41 catatan_dari_pembeli
42 catatan
43 userna

### Rename working columns

In [392]:
df_clean = df_clean.rename(columns={
    "no_pesanan": "order_id",
    "status_pesanan": "order_status",
    "alasan_pembatalan": "cancellation_reason",
    "status_pembatalan_pengembalian": "return_status",
    "no_resi": "tracking_number",
    "waktu_pesanan_dibuat": "order_created_at",
    "waktu_pembayaran_dilakukan": "payment_at",
    "waktu_pengiriman_diatur": "shipping_arranged_at",
    "waktu_pesanan_selesai": "order_completed_at",
    "nama_produk": "product_name",
    "sku_induk": "parent_sku",
    "nama_variasi": "variation_name",
    "jumlah": "quantity",
    "returned_quantity": "returned_quantity",
    "harga_awal": "original_price",
    "harga_setelah_diskon": "discounted_price",
    "subtotal_pesanan": "order_subtotal",
    "total_pembayaran": "total_payment"
})

In [393]:
df_clean.columns.tolist()

['order_id',
 'order_status',
 'cancellation_reason',
 'return_status',
 'tracking_number',
 'opsi_pengiriman',
 'antar_ke_counter_pick-up',
 'pesanan_harus_dikirimkan_sebelum_menghindari_keterlambatan',
 'shipping_arranged_at',
 'order_created_at',
 'payment_at',
 'tipe_pesanan',
 'metode_pembayaran',
 'parent_sku',
 'product_name',
 'nomor_referensi_sku',
 'variation_name',
 'original_price',
 'discounted_price',
 'quantity',
 'returned_quantity',
 'order_subtotal',
 'total_diskon',
 'diskon_dari_penjual',
 'diskon_dari_shopee',
 'berat_produk',
 'jumlah_produk_di_pesan',
 'total_berat',
 'voucher_ditanggung_penjual',
 'cashback_koin',
 'voucher_ditanggung_shopee',
 'paket_diskon',
 'paket_diskon_diskon_dari_shopee',
 'paket_diskon_diskon_dari_penjual',
 'potongan_koin_shopee',
 'diskon_kartu_kredit',
 'ongkos_kirim_dibayar_oleh_pembeli',
 'estimasi_potongan_biaya_pengiriman',
 'ongkos_kirim_pengembalian_barang',
 'total_payment',
 'perkiraan_ongkos_kirim',
 'catatan_dari_pembeli',
 

### Rename shipping deadline

In [394]:
df_clean = df_clean.rename(columns={
    "pesanan_harus_dikirimkan_sebelum_menghindari_keterlambatan":
        "shipping_deadline"
})

In [395]:
[col for col in df_clean.columns
if "shipping" in col or "order_" in col]

['order_id',
 'order_status',
 'shipping_deadline',
 'shipping_arranged_at',
 'order_created_at',
 'order_subtotal',
 'order_completed_at']

### Convert datetime

In [396]:
datetime_columns_clean = [
    "shipping_deadline",
    "shipping_arranged_at",
    "order_created_at",
    "payment_at",
    "order_completed_at"
]

In [397]:
for col in datetime_columns_clean:
  df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")


In [398]:
df_clean[datetime_columns_clean].dtypes

,0
shipping_deadline,datetime64[ns]
shipping_arranged_at,datetime64[ns]
order_created_at,datetime64[ns]
payment_at,datetime64[ns]
order_completed_at,datetime64[ns]


### Validation after conversion

because `errors="coerce"` can convert problematic strings into `NaT`.

In [399]:
df_clean[datetime_columns_clean].isna().sum()

,0
shipping_deadline,123
shipping_arranged_at,448
order_created_at,0
payment_at,120
order_completed_at,511


### Check the data period.

In [400]:
print (
    'Earliest order:',
    df_clean['order_created_at'].min()
)

Earliest order: 2025-10-01 00:00:00


In [401]:
print (
    'Latest order:',
    df_clean['order_created_at'].max()
)

Latest order: 2025-10-31 23:18:00


In [402]:
print(
    "Unique order dates:",
    df_clean["order_created_at"].dt.date.nunique()
)

Unique order dates: 31


In [403]:
df_clean["order_date"] = (
    df_clean["order_created_at"].dt.date
)

### Categorical consistency

In [404]:
df_clean["order_status"].value_counts(dropna=False)

,count
order_status,
Selesai,2755
Batal,511


In [405]:
df_clean["return_status"].value_counts(dropna=False)

,count
return_status,
NaN,3230
Permintaan Disetujui,35
Masalah Diselesaikan,1


In [406]:
df_clean["cancellation_reason"].value_counts(dropna=False)

,count
cancellation_reason,
NaN,2755
Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,124
Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada,112
Dibatalkan oleh Pembeli. Alasan: Need to change delivery address,85
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar,77
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal,61
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman,22
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan,10
Dibatalkan oleh Pembeli. Alasan: Penjual tidak responsif terhadap pertanyaan Pembeli,5


### Numeric validation

Quantity <= 0

In [407]:
df_clean[df_clean['quantity'] <= 0
][['order_id', 'product_name', 'quantity']
]

,order_id,product_name,quantity


Negative returned quantity

In [408]:
df_clean[df_clean['returned_quantity'] < 0
][['order_id', 'product_name', 'quantity', 'returned_quantity']
]

,order_id,product_name,quantity,returned_quantity


Returned quantity > ordered quantity

In [409]:
df_clean[df_clean['returned_quantity'] > df_clean['quantity']
][['order_id', 'product_name', 'quantity', 'returned_quantity']
]

,order_id,product_name,quantity,returned_quantity


Negative prices

In [410]:
df_clean[
    (df_clean['original_price'] < 0)
    |
    (df_clean['discounted_price'] < 0)
][['order_id', 'product_name', 'original_price','discounted_price']
]

,order_id,product_name,original_price,discounted_price


### Return consistency check

In [411]:
df_tes = pd.crosstab(
    df_clean["return_status"].fillna("No Return Status"),
    df_clean["returned_quantity"]
)

In [412]:
df_tes

returned_quantity,0,1
return_status,,
Masalah Diselesaikan,0,1
No Return Status,3230,0
Permintaan Disetujui,0,35


Returned quantity > 0 tetapi tidak ada return status

In [413]:
df_clean[
    (df_clean["returned_quantity"] > 0)
    &
    (df_clean["return_status"].isna())
][
    [
        "order_id",
        "return_status",
        "product_name",
        "quantity",
        "returned_quantity"
    ]
]

,order_id,return_status,product_name,quantity,returned_quantity


Ada return status tetapi returned quantity = 0

In [414]:
df_clean[
    (df_clean["return_status"].notna())
    &
    (df_clean["returned_quantity"] == 0)
][
    [
        "order_id",
        "return_status",
        "product_name",
        "quantity",
        "returned_quantity"
    ]
]

,order_id,return_status,product_name,quantity,returned_quantity


This is a real reconciliation check, not merely a cleanup.

### Return Data Validation

Return information was validated by cross-checking `return_status` against `returned_quantity`.

All 36 rows with a positive returned quantity also contain a return status:
- 35 rows are recorded as `Permintaan Disetujui`
- 1 row is recorded as `Masalah Diselesaikan`

No records were found where:
- returned quantity was positive without a return status, or
- a return status existed while returned quantity was zero.

Therefore, `returned_quantity > 0` can be used as the operational indicator for an actual returned item in subsequent analysis.

### Investigasi payment timestamp

In [415]:
payment_conversion_issue = df_clean[
    df_clean["payment_at"].isna()
].copy()

print(
    "Rows with invalid/missing payment timestamp:",
    len(payment_conversion_issue)
)

Rows with invalid/missing payment timestamp: 120


In [416]:
df_order.loc[
    payment_conversion_issue.index,
    [
        "No. Pesanan",
        "Status Pesanan",
        "Waktu Pesanan Dibuat",
        "Waktu Pembayaran Dilakukan",
        "Alasan Pembatalan"
    ]
].head(20)

,No. Pesanan,Status Pesanan,Waktu Pesanan Dibuat,Waktu Pembayaran Dilakukan,Alasan Pembatalan
9,251001V93P6B3G,Batal,2025-10-01 05:51,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
101,2510021MUKRH49,Batal,2025-10-02 05:26,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
111,2510021VWY57YN,Batal,2025-10-02 07:50,-,Dibatalkan secara otomatis oleh sistem Shopee....
118,25100221N8WWX3,Batal,2025-10-02 09:15,-,Dibatalkan secara otomatis oleh sistem Shopee....
164,2510023605996G,Batal,2025-10-02 20:05,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
170,2510023D7925C0,Batal,2025-10-02 22:14,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
223,25100351D96MEN,Batal,2025-10-03 13:48,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
227,2510035467V6E8,Batal,2025-10-03 14:38,-,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...
262,2510035RNWPBGY,Batal,2025-10-03 20:45,-,Dibatalkan oleh Pembeli. Alasan: Tidak ingin m...
265,2510035U1UK32W,Batal,2025-10-03 21:27,-,Dibatalkan secara otomatis oleh sistem Shopee....


### Unique raw values yang gagal dikonversi

In [417]:
df_order.loc[
    payment_conversion_issue.index,
    "Waktu Pembayaran Dilakukan"
].value_counts(dropna=False)

,count
Waktu Pembayaran Dilakukan,
-,120


### Link to order status

In [418]:
payment_conversion_issue[
    "order_status"
].value_counts(dropna=False)

,count
order_status,
Batal,120


In [419]:
payment_conversion_issue[
    "cancellation_reason"
].value_counts(dropna=False)

,count
cancellation_reason,
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar,77
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman,22
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan,10
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah Voucher,3
Dibatalkan oleh Pembeli. Alasan: Proses pembayaran sulit,3
Dibatalkan oleh Pembeli. Alasan: Tidak ingin membeli lagi,2
Dibatalkan oleh Pembeli. Alasan: Lainnya,2
Dibatalkan oleh Pembeli. Alasan: Menemukan yang lebih murah,1


### Investigation into missed shipping deadline

In [420]:
df_clean[
    df_clean["shipping_deadline"].isna()
][
    "order_status"
].value_counts(dropna=False)

,count
order_status,
Batal,123


In [421]:
df_clean[
    df_clean["shipping_deadline"].isna()
][
    "cancellation_reason"
].value_counts(dropna=False)

,count
cancellation_reason,
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar,77
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman,22
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan,10
Dibatalkan oleh Pembeli. Alasan: Perlu mengubah Voucher,3
Dibatalkan oleh Pembeli. Alasan: Proses pembayaran sulit,3
Dibatalkan oleh Pembeli. Alasan: Tidak ingin membeli lagi,2
Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,2
Dibatalkan oleh Pembeli. Alasan: Lainnya,2
Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada,1


### Cancelled orders yang sudah masuk fulfillment

In [422]:
cancelled_with_shipping = df_clean[
    (df_clean["order_status"] == "Batal")
    &
    (df_clean["shipping_arranged_at"].notna())
].copy()

print(
    "Cancelled rows with shipping arranged:",
    len(cancelled_with_shipping)
)

Cancelled rows with shipping arranged: 63


In [423]:
cancelled_with_shipping[
    [
        "order_id",
        "order_created_at",
        "payment_at",
        "shipping_arranged_at",
        "tracking_number",
        "cancellation_reason"
    ]
].head(20)

,order_id,order_created_at,payment_at,shipping_arranged_at,tracking_number,cancellation_reason
128,2510022A37A7M4,2025-10-02 11:46:00,2025-10-02 11:46:00,2025-10-02 20:29:00,SPXID05909168143A,Dibatalkan secara otomatis oleh sistem Shopee....
150,2510022XDJDQVE,2025-10-02 17:49:00,2025-10-02 17:49:00,2025-10-03 17:33:00,SPXID05815187860A,Dibatalkan secara otomatis oleh sistem Shopee....
448,251005ANA0P90S,2025-10-05 19:28:00,2025-10-05 19:28:00,2025-10-06 21:30:00,SPXID05136218692A,Dibatalkan secara otomatis oleh sistem Shopee....
472,251005AXY8YS1Q,2025-10-05 22:21:00,2025-10-05 22:21:00,2025-10-06 21:29:00,SPXID05353533556A,Dibatalkan secara otomatis oleh sistem Shopee....
473,251005AY44EGX8,2025-10-05 22:23:00,2025-10-05 22:23:00,2025-10-06 17:59:00,SPXID05633330548A,Dibatalkan secara otomatis oleh sistem Shopee....
485,251006B774TX09,2025-10-06 00:48:00,2025-10-06 00:48:00,2025-10-06 21:30:00,SPXID05449293962A,Dibatalkan secara otomatis oleh sistem Shopee....
510,251006C4800AUS,2025-10-06 09:28:00,2025-10-06 09:28:00,2025-10-06 18:04:00,SPXID05571504344A,Dibatalkan secara otomatis oleh sistem Shopee....
514,251006C570Q033,2025-10-06 09:45:00,2025-10-06 09:45:00,2025-10-07 18:46:00,SPXID05878964559A,Dibatalkan secara otomatis oleh sistem Shopee....
528,251006CAKFNNA7,2025-10-06 11:21:00,2025-10-06 11:21:00,2025-10-06 18:06:00,SPXID05732783134A,Dibatalkan secara otomatis oleh sistem Shopee....
534,251006CD6EP7WS,2025-10-06 12:08:00,2025-10-06 12:08:00,2025-10-07 13:46:00,SPXID05078393824A,Dibatalkan secara otomatis oleh sistem Shopee....


In [424]:
cancelled_with_shipping[
    "cancellation_reason"
].value_counts(dropna=False)

,count
cancellation_reason,
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal,61
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Paket hilang di perjalanan. Kompensasi yang memenuhi syarat telah dikreditkan ke Saldo Penjual-mu.,2


### Results of the additional investigation

First, the 120 `payment_at` entries that converted to `NaT` turned out not to be parsing errors; the raw values ​​were simply "-". All 120 of these rows were marked as "Cancelled." The primary cause—accounting for 77 rows—was that the orders had not yet been paid for. Thus, `NaT` is an appropriate representation for a payment timestamp that does not exist.

Second, all 123 missing `shipping_deadline` entries also originated from cancelled rows. The pattern closely mirrors that of orders lacking a payment timestamp.

Third, we identified 63 cancelled rows that already had a `shipping_arranged_at` timestamp. This is not a data error: 61 cases involved failed shipments, and 2 involved packages lost in transit. This indicates that these orders had already entered the fulfillment or logistics stage before ultimately being cancelled.

# Feature Engineering

New operational features are created from the cleaned dataset to support fulfillment performance, cancellation-stage, return, and bottleneck analysis.

The engineered features include:

- Payment lead time
- Processing time
- Order completion time
- Cancellation indicator
- Return indicator
- Fulfillment-start indicator
- Cancellation stage
- Late shipment indicator

### Payment Lead Time

In [425]:
df_clean["payment_lead_time_minutes"] = (
    df_clean["payment_at"]
    - df_clean["order_created_at"]
).dt.total_seconds() / 60

In [426]:
df_clean[
    [
        "order_created_at",
        "payment_at",
        "payment_lead_time_minutes"
    ]
].head(10)

,order_created_at,payment_at,payment_lead_time_minutes
0,2025-10-01 00:00:00,2025-10-01 00:01:00,1.0
1,2025-10-01 00:12:00,2025-10-01 00:12:00,0.0
2,2025-10-01 00:23:00,2025-10-01 00:23:00,0.0
3,2025-10-01 00:35:00,2025-10-01 00:35:00,0.0
4,2025-10-01 00:55:00,2025-10-01 00:55:00,0.0
5,2025-10-01 02:13:00,2025-10-01 02:15:00,2.0
6,2025-10-01 05:27:00,2025-10-01 14:42:00,555.0
7,2025-10-01 05:37:00,2025-10-01 05:37:00,0.0
8,2025-10-01 05:37:00,2025-10-01 05:37:00,0.0
9,2025-10-01 05:51:00,NaT,NaN


In [427]:
df_clean["payment_lead_time_minutes"].describe()

,payment_lead_time_minutes
count,3146.000000
mean,6.742530
std,73.661796
min,0.000000
25%,0.000000
50%,0.000000
75%,1.000000
max,1383.000000


In [428]:
df_clean[
    df_clean["payment_lead_time_minutes"] < 0
][
    [
        "order_id",
        "order_created_at",
        "payment_at",
        "payment_lead_time_minutes"
    ]
]

,order_id,order_created_at,payment_at,payment_lead_time_minutes


### Processing Time

In [429]:
df_clean["processing_time_minutes"] = (
    df_clean["shipping_arranged_at"]
    - df_clean["order_created_at"]
).dt.total_seconds() / 60

In [430]:
df_clean["processing_time_minutes"].describe()

,processing_time_minutes
count,2818.000000
mean,1103.298439
std,749.065279
min,0.000000
25%,479.000000
50%,1130.000000
75%,1489.750000
max,4766.000000


### Processing time percentiles.

In [431]:
df_clean["processing_time_minutes"].describe(
    percentiles=[
        0.50,
        0.75,
        0.90,
        0.95,
        0.99
    ]
)

,processing_time_minutes
count,2818.000000
mean,1103.298439
std,749.065279
min,0.000000
50%,1130.000000
75%,1489.750000
90%,1905.300000
95%,2673.850000
99%,3153.150000
max,4766.000000


### Completion Time

In [432]:
df_clean["completion_time_hours"] = (
    df_clean["order_completed_at"]
    - df_clean["shipping_arranged_at"]
).dt.total_seconds() / 3600

In [433]:
df_clean["completion_time_hours"].describe(
    percentiles=[
        0.50,
        0.75,
        0.90,
        0.95,
        0.99
    ]
)

,completion_time_hours
count,2755.000000
mean,182.236225
std,105.538879
min,4.816667
50%,191.450000
75%,240.783333
90%,310.603333
95%,366.365000
99%,480.367667
max,908.933333


### Completion Time Interpretation

Completion time measures the duration between shipping arrangement and final order completion.

Because this period may include logistics, delivery, and marketplace confirmation processes, it should not be interpreted purely as warehouse processing time.

### Cancellation Flag

In [434]:
df_clean["is_cancelled"] = (
    df_clean["order_status"] == "Batal"
).astype(int)

In [435]:
df_clean["is_cancelled"].value_counts()

,count
is_cancelled,
0,2755
1,511


### Return Flag

In [436]:
df_clean["is_returned"] = (
    df_clean["returned_quantity"] > 0
).astype(int)

In [437]:
df_clean["is_returned"].value_counts()

,count
is_returned,
0,3230
1,36


### Fulfillment Started Flag

In [438]:
df_clean["has_started_fulfillment"] = (
    df_clean["shipping_arranged_at"].notna()
).astype(int)

In [439]:
pd.crosstab(
    df_clean["order_status"],
    df_clean["has_started_fulfillment"]
)

has_started_fulfillment,0,1
order_status,,
Batal,448,63
Selesai,0,2755


### Cancellation Stage

In [440]:
def classify_cancellation_stage(row):
    if row["order_status"] != "Batal":
        return "Not Cancelled"

    elif pd.isna(row["shipping_arranged_at"]):
        return "Pre-Fulfillment"

    else:
        return "Post-Fulfillment Initiation"

In [441]:
df_clean["cancellation_stage"] = (
    df_clean.apply(
        classify_cancellation_stage,
        axis=1
    )
)

In [442]:
df_clean[
    "cancellation_stage"
].value_counts()

,count
cancellation_stage,
Not Cancelled,2755
Pre-Fulfillment,448
Post-Fulfillment Initiation,63


### Late Shipment

In [443]:
df_clean["is_late_shipment"] = (
    df_clean["shipping_arranged_at"]
    >
    df_clean["shipping_deadline"]
).astype(int)

In [444]:
df_clean["is_late_shipment"].value_counts()

,count
is_late_shipment,
0,3256
1,10


In [445]:
late_shipments = df_clean[
    df_clean["is_late_shipment"] == 1
].copy()

print(
    "Late shipment rows:",
    len(late_shipments)
)

Late shipment rows: 10


In [446]:
late_shipments[
    [
        "order_id",
        "order_status",
        "order_created_at",
        "shipping_deadline",
        "shipping_arranged_at",
        "processing_time_minutes"
    ]
].sort_values(
    "processing_time_minutes",
    ascending=False
).head(20)

,order_id,order_status,order_created_at,shipping_deadline,shipping_arranged_at,processing_time_minutes
2367,251023TEYU4P25,Selesai,2025-10-23 16:04:00,2025-10-24 23:59:00,2025-10-25 18:14:00,3010.0
572,251006D0PBUWQ7,Selesai,2025-10-06 17:57:00,2025-10-07 23:59:00,2025-10-08 18:19:00,2902.0
586,251006D3KNTG53,Batal,2025-10-06 18:49:00,2025-10-07 23:59:00,2025-10-08 18:19:00,2850.0
2104,251020K93E7FK2,Selesai,2025-10-20 19:30:00,2025-10-21 23:59:00,2025-10-22 17:15:00,2745.0
2426,251024UK40WN96,Selesai,2025-10-24 02:50:00,2025-10-24 23:59:00,2025-10-25 18:14:00,2364.0
982,251010P43WWNN2,Selesai,2025-10-10 08:52:00,2025-10-10 23:59:00,2025-10-11 18:18:00,2006.0
514,251006C570Q033,Batal,2025-10-06 09:45:00,2025-10-06 23:59:00,2025-10-07 18:46:00,1981.0
2992,251029BH9CV3CW,Selesai,2025-10-29 07:22:00,2025-10-29 23:59:00,2025-10-30 14:04:00,1842.0
2993,251029BH9CV3CW,Selesai,2025-10-29 07:22:00,2025-10-29 23:59:00,2025-10-30 14:04:00,1842.0
735,251008GC0QBFMG,Selesai,2025-10-08 01:57:00,2025-10-08 12:00:00,2025-10-08 12:13:00,616.0


### Deadline Margin

In [447]:
df_clean["deadline_margin_minutes"] = (
    df_clean["shipping_deadline"]
    - df_clean["shipping_arranged_at"]
).dt.total_seconds() / 60

In [448]:
df_clean["deadline_margin_minutes"].describe(
    percentiles=[0.01, 0.05, 0.10, 0.50]
)

,deadline_margin_minutes
count,2818.000000
mean,1108.101845
std,1266.746362
min,-1127.000000
1%,148.170000
5%,252.250000
10%,312.000000
50%,526.000000
max,5463.000000


- +120 → shipping arranged 120 menit sebelum deadline
- +10 → hanya 10 menit sebelum deadline
-   0 → tepat deadline
-  -30 → 30 menit melewati deadline

### Cross-check Cancellation Stage

In [449]:
pd.crosstab(
    df_clean["cancellation_stage"],
    df_clean["cancellation_reason"]
)

cancellation_reason,Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,Dibatalkan oleh Pembeli. Alasan: Lainnya,Dibatalkan oleh Pembeli. Alasan: Menemukan yang lebih murah,Dibatalkan oleh Pembeli. Alasan: Need to change delivery address,Dibatalkan oleh Pembeli. Alasan: Penjual tidak responsif terhadap pertanyaan Pembeli,Dibatalkan oleh Pembeli. Alasan: Perlu mengubah Voucher,Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman,Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan,Dibatalkan oleh Pembeli. Alasan: Proses pembayaran sulit,Dibatalkan oleh Pembeli. Alasan: Tidak ingin membeli lagi,Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada,Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Paket hilang di perjalanan. Kompensasi yang memenuhi syarat telah dikreditkan ke Saldo Penjual-mu.,Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal,Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Penjual gagal mengirimkan pesanan tepat waktu,Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Penjual tidak mengatur pengiriman tepat waktu,Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar
cancellation_stage,,,,,,,,,,,,,,,,
Post-Fulfillment Initiation,0,0,0,0,0,0,0,0,0,0,0,2,61,0,0,0
Pre-Fulfillment,124,2,1,85,5,3,22,10,3,2,112,0,0,1,1,77


### Processing Category

Since the actual distribution is:
- P90 = 1905.30 minutes
- P95 = 2673.85 minutes
- P99 = 3153.15 minutes

So, we use the percentile as the operational monitoring threshold:

### Processing Time Classification

Processing-time categories are defined using the empirical distribution of the dataset:

- Normal: ≤ 90th percentile
- Attention: > P90 to P95
- Slow: > P95 to P99
- Critical: > P99

These categories are intended for operational monitoring and prioritization rather than representing predefined service-level agreements.

In [450]:
p90_processing = df_clean[
    "processing_time_minutes"
].quantile(0.90)

p95_processing = df_clean[
    "processing_time_minutes"
].quantile(0.95)

p99_processing = df_clean[
    "processing_time_minutes"
].quantile(0.99)

print("P90:", p90_processing)
print("P95:", p95_processing)
print("P99:", p99_processing)

P90: 1905.3000000000002
P95: 2673.8500000000017
P99: 3153.1499999999996


In [451]:
def classify_processing_time(x):
    if pd.isna(x):
        return "Not Processed"
    elif x <= p90_processing:
        return "Normal"
    elif x <= p95_processing:
        return "Attention"
    elif x <= p99_processing:
        return "Slow"
    else:
        return "Critical"

df_clean["processing_category"] = (
    df_clean["processing_time_minutes"]
    .apply(classify_processing_time)
)

In [452]:
df_clean[
    "processing_category"
].value_counts()

,count
processing_category,
Normal,2536
Not Processed,448
Attention,141
Slow,112
Critical,29


In [453]:
df_clean[
    "processing_category"
].value_counts(
    normalize=True
).mul(100).round(2)

,proportion
processing_category,
Normal,77.65
Not Processed,13.72
Attention,4.32
Slow,3.43
Critical,0.89


### Critical Processing Investigation

In [454]:
critical_processing = df_clean[
    df_clean["processing_category"] == "Critical"
].copy()

print(
    "Critical processing rows:",
    len(critical_processing)
)

Critical processing rows: 29


In [455]:
critical_processing[
    [
        "order_id",
        "order_status",
        "order_created_at",
        "shipping_arranged_at",
        "shipping_deadline",
        "processing_time_minutes",
        "deadline_margin_minutes",
        "product_name"
    ]
].sort_values(
    "processing_time_minutes",
    ascending=False
).head(20)

,order_id,order_status,order_created_at,shipping_arranged_at,shipping_deadline,processing_time_minutes,deadline_margin_minutes,product_name
226,251003529YJKXG,Selesai,2025-10-03 14:04:00,2025-10-06 21:30:00,2025-10-06 23:59:00,4766.0,149.0,Tongsis Tripod HP 3 in 1 dengan Lampu LED - VI...
3229,251031HBT9PC8Q,Selesai,2025-10-31 15:00:00,2025-11-03 18:04:00,2025-11-03 23:59:00,4504.0,355.0,VIDRONA ELK-113 Lampu LED 11 inch Tripod Ring ...
257,2510035NQTX00X,Selesai,2025-10-03 19:52:00,2025-10-06 21:30:00,2025-10-06 23:59:00,4418.0,149.0,Tongsis Tripod HP 3 in 1 dengan Lampu LED - VI...
3246,251031HSSJDJXA,Selesai,2025-10-31 19:10:00,2025-11-03 18:08:00,2025-11-03 23:59:00,4258.0,351.0,Tambahan Packing Dengan Bubble Wrap Supaya Pak...
3247,251031HSSJDJXA,Selesai,2025-10-31 19:10:00,2025-11-03 18:08:00,2025-11-03 23:59:00,4258.0,351.0,VIDRONA ELK-113 Lampu LED 11 inch Tripod Ring ...
268,2510035Y8N4NNA,Selesai,2025-10-03 22:42:00,2025-10-06 18:03:00,2025-10-06 23:59:00,4041.0,356.0,VIDRONA AKS-156 Stand X Banner 60x160 cm Rangk...
275,25100463D8080U,Selesai,2025-10-03 23:57:00,2025-10-06 18:06:00,2025-10-06 23:59:00,3969.0,353.0,VIDRONA Sarung Tripod Tas Tripod 2 Meter Tas L...
273,25100463D8080U,Selesai,2025-10-03 23:57:00,2025-10-06 18:06:00,2025-10-06 23:59:00,3969.0,353.0,Tambahan Packing Dengan Bubble Wrap Supaya Pak...
274,25100463D8080U,Selesai,2025-10-03 23:57:00,2025-10-06 18:06:00,2025-10-06 23:59:00,3969.0,353.0,VIDRONA Tripod 2M + Ringlight 26 Cm Remot Blue...
1852,251018CFGXWMW9,Selesai,2025-10-18 02:36:00,2025-10-20 18:06:00,2025-10-20 23:59:00,3810.0,353.0,VIDRONA Maca 6 Pro Bluetooth Headset Earphone ...


### Late Shipment Investigation

Found 10 late-shipment rows, since we know the dataset is at the item level.

This means we cannot yet say:

10 late orders.

The correct term at the moment is:

10 late-shipment item rows.

In [456]:
print(
    "Late shipment rows:",
    len(late_shipments)
)

print(
    "Unique late-shipment orders:",
    late_shipments["order_id"].nunique()
)

Late shipment rows: 10
Unique late-shipment orders: 9


In [457]:
late_shipments[
    "order_status"
].value_counts()

,count
order_status,
Selesai,8
Batal,2


In [458]:
late_shipments[
    "cancellation_reason"
].value_counts(dropna=False)

,count
cancellation_reason,
NaN,8
Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal,2


### Validation of multi-item order consistency

In [459]:
status_consistency = (
    df_clean
    .groupby("order_id")["order_status"]
    .nunique()
)

print(
    "Orders with multiple order statuses:",
    (status_consistency > 1).sum()
)

Orders with multiple order statuses: 0


Cancellation stage

In [460]:
cancellation_stage_consistency = (
    df_clean
    .groupby("order_id")[
        "cancellation_stage"
    ]
    .nunique()
)

print(
    "Orders with multiple cancellation stages:",
    (cancellation_stage_consistency > 1).sum()
)

Orders with multiple cancellation stages: 0


Shipping timestamp

In [461]:
shipping_consistency = (
    df_clean
    .groupby("order_id")[
        "shipping_arranged_at"
    ]
    .nunique(dropna=False)
)

print(
    "Orders with multiple shipping timestamps:",
    (shipping_consistency > 1).sum()
)

Orders with multiple shipping timestamps: 0


Processing category

In [462]:
processing_consistency = (
    df_clean
    .groupby("order_id")[
        "processing_category"
    ]
    .nunique()
)

print(
    "Orders with multiple processing categories:",
    (processing_consistency > 1).sum()
)

Orders with multiple processing categories: 0


They are all 0, we have evidence that the attribute is indeed consistent across orders containing multiple items.

### Returns require different handling.

A return should not immediately be considered an order-level attribute.

Order A
- Product 1 → returned
- Product 2 → not returned
- Product 3 → not returned

In [463]:
return_consistency = (
    df_clean
    .groupby("order_id")[
        "is_returned"
    ]
    .nunique()
)

print(
    "Orders containing mixed return status across items:",
    (return_consistency > 1).sum()
)

Orders containing mixed return status across items: 1


this 1 order might means some items in that order are being returned.

## Order-Level Operational Table

The raw dataset is stored at the order-item level. A separate order-level table is therefore created to calculate operational KPIs without double-counting multi-item orders.

Order-level operational fields are aggregated carefully, while item-level return information is summarized across all items within each order.

In [464]:
df_orders = (
    df_clean
    .groupby("order_id", as_index=False)
    .agg(
        order_status=("order_status", "first"),
        order_created_at=("order_created_at", "first"),
        order_date=("order_date", "first"),
        payment_at=("payment_at", "first"),
        shipping_deadline=("shipping_deadline", "first"),
        shipping_arranged_at=("shipping_arranged_at", "first"),
        order_completed_at=("order_completed_at", "first"),

        payment_lead_time_minutes=(
            "payment_lead_time_minutes",
            "first"
        ),

        processing_time_minutes=(
            "processing_time_minutes",
            "first"
        ),

        completion_time_hours=(
            "completion_time_hours",
            "first"
        ),

        deadline_margin_minutes=(
            "deadline_margin_minutes",
            "first"
        ),

        is_cancelled=("is_cancelled", "max"),

        has_started_fulfillment=(
            "has_started_fulfillment",
            "max"
        ),

        cancellation_stage=(
            "cancellation_stage",
            "first"
        ),

        processing_category=(
            "processing_category",
            "first"
        ),

        is_late_shipment=(
            "is_late_shipment",
            "max"
        ),

        is_returned=(
            "is_returned",
            "max"
        ),

        returned_items=(
            "returned_quantity",
            "sum"
        ),

        total_items=(
            "quantity",
            "sum"
        ),

        unique_products=(
            "product_name",
            "nunique"
        )
    )
)

In [465]:
len(df_orders)

3153

In [466]:
df_orders.head()

,order_id,order_status,order_created_at,order_date,payment_at,shipping_deadline,shipping_arranged_at,order_completed_at,payment_lead_time_minutes,processing_time_minutes,...,deadline_margin_minutes,is_cancelled,has_started_fulfillment,cancellation_stage,processing_category,is_late_shipment,is_returned,returned_items,total_items,unique_products
0,251001004TP8HH,Selesai,2025-10-01 13:42:00,2025-10-01,2025-10-01 13:43:00,2025-10-02 23:59:00,2025-10-02 18:35:00,2025-10-13 08:45:00,1.0,1733.0,...,324.0,0,1,Not Cancelled,Normal,0,0,0,1,1
1,25100100BXNHW4,Selesai,2025-10-01 13:46:00,2025-10-01,2025-10-01 13:46:00,2025-10-02 23:59:00,2025-10-02 18:36:00,2025-10-03 19:41:00,0.0,1730.0,...,323.0,0,1,Not Cancelled,Normal,0,0,0,1,1
2,25100100DK286F,Selesai,2025-10-01 13:47:00,2025-10-01,2025-10-01 13:48:00,2025-10-02 23:59:00,2025-10-02 18:36:00,2025-10-11 19:50:00,1.0,1729.0,...,323.0,0,1,Not Cancelled,Normal,0,0,0,1,1
3,25100100J431RT,Batal,2025-10-01 13:50:00,2025-10-01,2025-10-01 13:50:00,2025-10-02 23:59:00,NaT,NaT,0.0,NaN,...,NaN,1,0,Pre-Fulfillment,Not Processed,0,0,0,1,1
4,25100101F6V097,Batal,2025-10-01 14:06:00,2025-10-01,2025-10-01 14:06:00,2025-10-02 23:59:00,NaT,NaT,0.0,NaN,...,NaN,1,0,Pre-Fulfillment,Not Processed,0,0,0,1,1


In [467]:
df_orders.columns

Index(['order_id', 'order_status', 'order_created_at', 'order_date',
       'payment_at', 'shipping_deadline', 'shipping_arranged_at',
       'order_completed_at', 'payment_lead_time_minutes',
       'processing_time_minutes', 'completion_time_hours',
       'deadline_margin_minutes', 'is_cancelled', 'has_started_fulfillment',
       'cancellation_stage', 'processing_category', 'is_late_shipment',
       'is_returned', 'returned_items', 'total_items', 'unique_products'],
      dtype='object')

### Order-Level Table Validation

In [468]:
print(
    "Item-level rows:",
    len(df_clean)
)

print(
    "Order-level rows:",
    len(df_orders)
)

print(
    "Unique order IDs:",
    df_orders["order_id"].nunique()
)

Item-level rows: 3266
Order-level rows: 3153
Unique order IDs: 3153


In [469]:
df_orders["order_id"].duplicated().sum()

np.int64(0)

### Sanity Check Aggregation

In [470]:
print(
    "Item-level total quantity:",
    df_clean["quantity"].sum()
)

print(
    "Order-level aggregated quantity:",
    df_orders["total_items"].sum()
)

Item-level total quantity: 3513
Order-level aggregated quantity: 3513


In [471]:
print(
    "Item-level returned quantity:",
    df_clean["returned_quantity"].sum()
)

print(
    "Order-level returned items:",
    df_orders["returned_items"].sum()
)

Item-level returned quantity: 36
Order-level returned items: 36


## Operational KPI Analysis

Operational KPIs are calculated using the order-level dataset (`df_orders`) to prevent multi-item orders from being double-counted.

The analysis focuses on:

- Order completion performance
- Cancellation performance
- Fulfillment processing performance
- Late shipment performance
- Return performance
- Cancellation stage

In [472]:
total_orders = len(df_orders)

completed_orders = (
    df_orders["order_status"] == "Selesai"
).sum()

cancelled_orders = (
    df_orders["order_status"] == "Batal"
).sum()

completion_rate = (
    completed_orders / total_orders * 100
)

cancellation_rate = (
    cancelled_orders / total_orders * 100
)

print(f"Total Orders      : {total_orders:,}")
print(f"Completed Orders  : {completed_orders:,}")
print(f"Cancelled Orders  : {cancelled_orders:,}")
print(f"Completion Rate   : {completion_rate:.2f}%")
print(f"Cancellation Rate : {cancellation_rate:.2f}%")

Total Orders      : 3,153
Completed Orders  : 2,650
Cancelled Orders  : 503
Completion Rate   : 84.05%
Cancellation Rate : 15.95%


### Processing KPI

In [473]:
processed_orders = df_orders[
    df_orders["shipping_arranged_at"].notna()
].copy()

print(
    "Processed Orders:",
    len(processed_orders)
)

Processed Orders: 2712


In [474]:
median_processing_time = (
    processed_orders["processing_time_minutes"].median()
)

p90_processing_time = (
    processed_orders["processing_time_minutes"].quantile(0.90)
)

p95_processing_time = (
    processed_orders["processing_time_minutes"].quantile(0.95)
)

p99_processing_time = (
    processed_orders["processing_time_minutes"].quantile(0.99)
)

print(
    f"Median Processing Time : "
    f"{median_processing_time:.2f} minutes"
)

print(
    f"P90 Processing Time    : "
    f"{p90_processing_time:.2f} minutes"
)

print(
    f"P95 Processing Time    : "
    f"{p95_processing_time:.2f} minutes"
)

print(
    f"P99 Processing Time    : "
    f"{p99_processing_time:.2f} minutes"
)

Median Processing Time : 1132.50 minutes
P90 Processing Time    : 1926.10 minutes
P95 Processing Time    : 2695.40 minutes
P99 Processing Time    : 3141.67 minutes


In [475]:
print(
    f"Median Processing Time : "
    f"{median_processing_time / 60:.2f} hours"
)

Median Processing Time : 18.88 hours


### Critical Order KPI

In [476]:
critical_orders = (
    df_orders["processing_category"] == "Critical"
).sum()

critical_order_rate = (
    critical_orders
    / len(processed_orders)
    * 100
)

print(f"Critical Orders     : {critical_orders:,}")
print(f"Critical Order Rate : {critical_order_rate:.2f}%")

Critical Orders     : 26
Critical Order Rate : 0.96%


### Late Shipment KPI

In [477]:
late_orders = (
    processed_orders["is_late_shipment"] == 1
).sum()

late_shipment_rate = (
    late_orders
    / len(processed_orders)
    * 100
)

on_time_rate = 100 - late_shipment_rate

print(f"Late Shipment Orders : {late_orders:,}")
print(f"Late Shipment Rate   : {late_shipment_rate:.2f}%")
print(f"On-Time Rate         : {on_time_rate:.2f}%")

Late Shipment Orders : 9
Late Shipment Rate   : 0.33%
On-Time Rate         : 99.67%


### Return KPI

In [478]:
returned_orders = (
    df_orders["is_returned"] == 1
).sum()

return_order_rate = (
    returned_orders
    / completed_orders
    * 100
)

print(f"Orders With Return : {returned_orders:,}")
print(f"Return Order Rate  : {return_order_rate:.2f}%")
print(
    f"Returned Items     : "
    f"{df_orders['returned_items'].sum():,.0f}"
)

Orders With Return : 36
Return Order Rate  : 1.36%
Returned Items     : 36


Return Order Rate ≠ Returned Item Rate.

In [479]:
returned_item_rate = (
    df_orders["returned_items"].sum()
    / df_orders["total_items"].sum()
    * 100
)

print(
    f"Returned Item Rate: "
    f"{returned_item_rate:.2f}%"
)

Returned Item Rate: 1.02%


Return Order Rate
- % order yang memiliki minimal satu return

Returned Item Rate
- % quantity barang yang diretur

### Cancellation Stage KPI

In [480]:
df_orders[
    "cancellation_stage"
].value_counts()

,count
cancellation_stage,
Not Cancelled,2650
Pre-Fulfillment,441
Post-Fulfillment Initiation,62


In [481]:
cancelled_order_df = df_orders[
    df_orders["is_cancelled"] == 1
].copy()

cancellation_stage_summary = (
    cancelled_order_df[
        "cancellation_stage"
    ]
    .value_counts()
    .reset_index()
)

cancellation_stage_summary.columns = [
    "cancellation_stage",
    "cancelled_orders"
]

cancellation_stage_summary[
    "percentage"
] = (
    cancellation_stage_summary["cancelled_orders"]
    / len(cancelled_order_df)
    * 100
).round(2)

cancellation_stage_summary

,cancellation_stage,cancelled_orders,percentage
0,Pre-Fulfillment,441,87.67
1,Post-Fulfillment Initiation,62,12.33


### Cancellation Reasons

In [482]:
cancellation_reason_summary = (
    df_clean[
        df_clean["order_status"] == "Batal"
    ]
    .groupby(
        [
            "order_id",
            "cancellation_reason",
            "cancellation_stage"
        ],
        dropna=False
    )
    .size()
    .reset_index()
)

In [483]:
cancellation_reason_summary = (
    cancellation_reason_summary[
        [
            "order_id",
            "cancellation_reason",
            "cancellation_stage"
        ]
    ]
    .drop_duplicates()
)

In [484]:
cancellation_reason_kpi = (
    cancellation_reason_summary
    .groupby(
        [
            "cancellation_reason",
            "cancellation_stage"
        ],
        dropna=False
    )
    .agg(
        cancelled_orders=("order_id", "count")
    )
    .reset_index()
    .sort_values(
        "cancelled_orders",
        ascending=False
    )
)

cancellation_reason_kpi.head(15)

,cancellation_reason,cancellation_stage,cancelled_orders
0,Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,Pre-Fulfillment,121
10,Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan ...,Pre-Fulfillment,111
3,Dibatalkan oleh Pembeli. Alasan: Need to chang...,Pre-Fulfillment,83
15,Dibatalkan secara otomatis oleh sistem Shopee....,Pre-Fulfillment,76
12,Dibatalkan secara otomatis oleh sistem Shopee....,Post-Fulfillment Initiation,60
6,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...,Pre-Fulfillment,22
7,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...,Pre-Fulfillment,10
4,Dibatalkan oleh Pembeli. Alasan: Penjual tidak...,Pre-Fulfillment,5
5,Dibatalkan oleh Pembeli. Alasan: Perlu menguba...,Pre-Fulfillment,3
8,Dibatalkan oleh Pembeli. Alasan: Proses pembay...,Pre-Fulfillment,3


In [485]:
cancellation_reason_kpi['cancellation_reason'].unique()

array(['Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya',
       'Dibatalkan oleh Pembeli. Alasan: Ubah Pesanan yang Ada',
       'Dibatalkan oleh Pembeli. Alasan: Need to change delivery address',
       'Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pesanan belum dibayar',
       'Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Pengiriman gagal',
       'Dibatalkan oleh Pembeli. Alasan: Perlu mengubah alamat pengiriman',
       'Dibatalkan oleh Pembeli. Alasan: Perlu mengubah pesanan',
       'Dibatalkan oleh Pembeli. Alasan: Penjual tidak responsif terhadap pertanyaan Pembeli',
       'Dibatalkan oleh Pembeli. Alasan: Perlu mengubah Voucher',
       'Dibatalkan oleh Pembeli. Alasan: Proses pembayaran sulit',
       'Dibatalkan secara otomatis oleh sistem Shopee. Alasan: Paket hilang di perjalanan. Kompensasi yang memenuhi syarat telah dikreditkan ke Saldo Penjual-mu.',
       'Dibatalkan oleh Pembeli. Alasan: Lainnya',
       'Dibatalkan oleh Pembeli. Alasan: Tidak i

### Deadline Risk

Since the median margin deadline is around 526 minutes (8.8 hours).

Now we create indicator for orders that are nearing their deadline, even though they are not yet late.

we use a 2-hour threshold:

In [486]:
df_orders["is_deadline_risk"] = (
    (df_orders["deadline_margin_minutes"] >= 0)
    &
    (df_orders["deadline_margin_minutes"] <= 120)
).astype(int)

In [487]:
deadline_risk_orders = (
    processed_orders[
        (
            processed_orders["deadline_margin_minutes"] >= 0
        )
        &
        (
            processed_orders["deadline_margin_minutes"] <= 120
        )
    ]
)

print(
    "Orders shipped within 2 hours of deadline:",
    len(deadline_risk_orders)
)

Orders shipped within 2 hours of deadline: 10


In [488]:
processed_orders = df_orders[
    df_orders["shipping_arranged_at"].notna()
].copy()

In [489]:
df_orders["is_deadline_risk"].value_counts()

,count
is_deadline_risk,
0,3143
1,10


### Daily Operational Monitoring

In [490]:
daily_operations = (
    df_orders
    .groupby("order_date")
    .agg(
        total_orders=("order_id", "count"),

        completed_orders=(
            "order_status",
            lambda x: (x == "Selesai").sum()
        ),

        cancelled_orders=(
            "is_cancelled",
            "sum"
        ),

        returned_orders=(
            "is_returned",
            "sum"
        ),

        late_orders=(
            "is_late_shipment",
            "sum"
        ),

        critical_orders=(
            "processing_category",
            lambda x: (x == "Critical").sum()
        ),

        median_processing_time=(
            "processing_time_minutes",
            "median"
        )
    )
    .reset_index()
)

In [491]:
# adding rates

daily_operations["completion_rate"] = (
    daily_operations["completed_orders"]
    / daily_operations["total_orders"]
    * 100
)

daily_operations["cancellation_rate"] = (
    daily_operations["cancelled_orders"]
    / daily_operations["total_orders"]
    * 100
)

daily_operations["late_shipment_rate"] = (
    daily_operations["late_orders"]
    / daily_operations["total_orders"]
    * 100
)

daily_operations.head(10)

,order_date,total_orders,completed_orders,cancelled_orders,returned_orders,late_orders,critical_orders,median_processing_time,completion_rate,cancellation_rate,late_shipment_rate
0,2025-10-01,94,81,13,2,0,0,792.0,86.170213,13.829787,0.000000
1,2025-10-02,78,64,14,0,0,0,680.5,82.051282,17.948718,0.000000
2,2025-10-03,93,81,12,1,0,4,1151.0,87.096774,12.903226,0.000000
3,2025-10-04,101,85,16,0,0,10,2671.0,84.158416,15.841584,0.000000
4,2025-10-05,99,83,16,2,0,0,1497.5,83.838384,16.161616,0.000000
5,2025-10-06,138,109,29,0,3,0,1109.0,78.985507,21.014493,2.173913
6,2025-10-07,101,83,18,0,0,0,1158.0,82.178218,17.821782,0.000000
7,2025-10-08,118,96,22,1,1,0,1180.0,81.355932,18.644068,0.847458
8,2025-10-09,86,75,11,0,0,0,1213.0,87.209302,12.790698,0.000000
9,2025-10-10,141,122,19,0,1,0,1109.0,86.524823,13.475177,0.709220


### The busiest day

In [492]:
daily_operations.sort_values(
    "total_orders",
    ascending=False
).head(10)

,order_date,total_orders,completed_orders,cancelled_orders,returned_orders,late_orders,critical_orders,median_processing_time,completion_rate,cancellation_rate,late_shipment_rate
9,2025-10-10,141,122,19,0,1,0,1109.0,86.524823,13.475177,0.709220
5,2025-10-06,138,109,29,0,3,0,1109.0,78.985507,21.014493,2.173913
12,2025-10-13,136,112,24,1,0,0,940.0,82.352941,17.647059,0.000000
24,2025-10-25,128,107,21,0,0,5,2603.0,83.593750,16.406250,0.000000
14,2025-10-15,126,102,24,1,0,0,990.0,80.952381,19.047619,0.000000
7,2025-10-08,118,96,22,1,1,0,1180.0,81.355932,18.644068,0.847458
23,2025-10-24,110,96,14,1,1,0,966.0,87.272727,12.727273,0.909091
26,2025-10-27,108,90,18,4,0,0,906.0,83.333333,16.666667,0.000000
22,2025-10-23,108,90,18,1,1,0,853.0,83.333333,16.666667,0.925926
10,2025-10-11,108,86,22,6,0,1,788.0,79.629630,20.370370,0.000000


Latest processing day:

In [493]:
daily_operations.sort_values(
    "median_processing_time",
    ascending=False
).head(10)

,order_date,total_orders,completed_orders,cancelled_orders,returned_orders,late_orders,critical_orders,median_processing_time,completion_rate,cancellation_rate,late_shipment_rate
3,2025-10-04,101,85,16,0,0,10,2671.0,84.158416,15.841584,0.000000
24,2025-10-25,128,107,21,0,0,5,2603.0,83.593750,16.406250,0.000000
17,2025-10-18,76,59,17,1,0,3,2423.0,77.631579,22.368421,0.000000
25,2025-10-26,102,85,17,1,0,0,1598.0,83.333333,16.666667,0.000000
4,2025-10-05,99,83,16,2,0,0,1497.5,83.838384,16.161616,0.000000
18,2025-10-19,93,82,11,1,0,0,1445.5,88.172043,11.827957,0.000000
11,2025-10-12,107,89,18,0,0,0,1426.0,83.177570,16.822430,0.000000
8,2025-10-09,86,75,11,0,0,0,1213.0,87.209302,12.790698,0.000000
21,2025-10-22,93,80,13,2,0,0,1197.5,86.021505,13.978495,0.000000
7,2025-10-08,118,96,22,1,1,0,1180.0,81.355932,18.644068,0.847458


### Workload vs Processing Time

In [494]:
volume_processing_correlation = (
    daily_operations["total_orders"]
    .corr(
        daily_operations[
            "median_processing_time"
        ]
    )
)

print(
    "Correlation between daily order volume "
    "and median processing time:",
    round(
        volume_processing_correlation,
        3
    )
)

Correlation between daily order volume and median processing time: 0.076


### Executive KPI Table

In [495]:
operational_kpi = pd.DataFrame({
    "metric": [
        "Total Orders",
        "Completed Orders",
        "Cancelled Orders",
        "Completion Rate (%)",
        "Cancellation Rate (%)",
        "Median Processing Time (Hours)",
        "P90 Processing Time (Hours)",
        "Critical Orders",
        "Critical Order Rate (%)",
        "Late Shipment Orders",
        "Late Shipment Rate (%)",
        "On-Time Shipping Arrangement Rate (%)",
        "Orders With Return",
        "Return Order Rate (%)",
        "Returned Items",
        "Returned Item Rate (%)"
    ],

    "value": [
        total_orders,
        completed_orders,
        cancelled_orders,
        round(completion_rate, 2),
        round(cancellation_rate, 2),
        round(median_processing_time / 60, 2),
        round(p90_processing_time / 60, 2),
        critical_orders,
        round(critical_order_rate, 2),
        late_orders,
        round(late_shipment_rate, 2),
        round(on_time_rate, 2),
        returned_orders,
        round(return_order_rate, 2),
        df_orders["returned_items"].sum(),
        round(returned_item_rate, 2)
    ]
})

operational_kpi

,metric,value
0,Total Orders,3153.00
1,Completed Orders,2650.00
2,Cancelled Orders,503.00
3,Completion Rate (%),84.05
4,Cancellation Rate (%),15.95
5,Median Processing Time (Hours),18.88
6,P90 Processing Time (Hours),32.10
7,Critical Orders,26.00
8,Critical Order Rate (%),0.96
9,Late Shipment Orders,9.00


## Final Reconciliation Checks

### Operational Reconciliation Checks

Operational reconciliation checks are performed at the order level to identify logical inconsistencies across order status, fulfillment, cancellation, return, and timestamp information.

Completed tetapi tidak ada shipping timestamp

In [496]:
check_completed_no_shipping = df_orders[
    (df_orders["order_status"] == "Selesai")
    &
    (df_orders["shipping_arranged_at"].isna())
]

len(check_completed_no_shipping)

0

Completed tetapi tidak ada completion timestamp

In [497]:
check_completed_no_completion = df_orders[
    (df_orders["order_status"] == "Selesai")
    &
    (df_orders["order_completed_at"].isna())
]

len(check_completed_no_completion)

0

Shipping sebelum order dibuat

In [498]:
check_shipping_before_order = df_orders[
    df_orders["shipping_arranged_at"]
    <
    df_orders["order_created_at"]
]

len(check_shipping_before_order)

0

Completion sebelum shipping arrangement

In [499]:
check_completion_before_shipping = df_orders[
    df_orders["order_completed_at"]
    <
    df_orders["shipping_arranged_at"]
]

len(check_completion_before_shipping)

0

Payment sebelum order dibuat

In [500]:
check_payment_before_order = df_orders[
    df_orders["payment_at"]
    <
    df_orders["order_created_at"]
]

len(check_payment_before_order)

0

### Return Reconciliation

In [501]:
check_return_without_items = df_orders[
    (df_orders["is_returned"] == 1)
    &
    (df_orders["returned_items"] <= 0)
]

len(check_return_without_items)

0

In [502]:
check_items_without_return_flag = df_orders[
    (df_orders["returned_items"] > 0)
    &
    (df_orders["is_returned"] == 0)
]

len(check_items_without_return_flag)

0

### Cancellation Stage Reconciliation

In [503]:
check_cancelled_wrong_stage = df_orders[
    (df_orders["is_cancelled"] == 1)
    &
    (df_orders["cancellation_stage"] == "Not Cancelled")
]

len(check_cancelled_wrong_stage)

0

In [504]:
check_completed_cancel_stage = df_orders[
    (df_orders["is_cancelled"] == 0)
    &
    (df_orders["cancellation_stage"] != "Not Cancelled")
]

len(check_completed_cancel_stage)

0

### Reconciliation Summary

In [505]:
reconciliation_summary = pd.DataFrame({
    "check": [
        "Completed order without shipping timestamp",
        "Completed order without completion timestamp",
        "Shipping before order creation",
        "Completion before shipping arrangement",
        "Payment before order creation",
        "Return flag without returned items",
        "Returned items without return flag",
        "Cancelled order classified as Not Cancelled",
        "Non-cancelled order with cancellation stage"
    ],

    "issue_count": [
        len(check_completed_no_shipping),
        len(check_completed_no_completion),
        len(check_shipping_before_order),
        len(check_completion_before_shipping),
        len(check_payment_before_order),
        len(check_return_without_items),
        len(check_items_without_return_flag),
        len(check_cancelled_wrong_stage),
        len(check_completed_cancel_stage)
    ]
})

In [506]:
reconciliation_summary["status"] = np.where(
    reconciliation_summary["issue_count"] == 0,
    "PASS",
    "REVIEW"
)

reconciliation_summary

,check,issue_count,status
0,Completed order without shipping timestamp,0,PASS
1,Completed order without completion timestamp,0,PASS
2,Shipping before order creation,0,PASS
3,Completion before shipping arrangement,0,PASS
4,Payment before order creation,0,PASS
5,Return flag without returned items,0,PASS
6,Returned items without return flag,0,PASS
7,Cancelled order classified as Not Cancelled,0,PASS
8,Non-cancelled order with cancellation stage,0,PASS


### Cancellation Reason Standardization

Currently, the cancellation reason appears as a long sentence from Shopee. That is not ideal for the dashboard.

In [507]:
def categorize_cancellation_reason(reason):

    if pd.isna(reason):
        return "Not Cancelled"

    reason_lower = reason.lower()

    if "belum dibayar" in reason_lower:
        return "Unpaid Order"

    elif "pengiriman gagal" in reason_lower:
        return "Delivery Failed"

    elif "paket hilang" in reason_lower:
        return "Package Lost"

    elif "alamat" in reason_lower or "delivery address" in reason_lower:
        return "Change Address"

    elif "ubah pesanan" in reason_lower or "mengubah pesanan" in reason_lower:
        return "Change Order"

    elif "voucher" in reason_lower:
        return "Voucher Issue"

    elif "tidak responsif" in reason_lower:
        return "Seller Responsiveness"

    elif "pembayaran sulit" in reason_lower:
        return "Payment Difficulty"

    elif "tidak ingin membeli" in reason_lower:
        return "Changed Mind"

    elif "lebih murah" in reason_lower:
        return "Found Cheaper Alternative"

    elif "lainnya" in reason_lower:
        return "Other"
    elif ("penjual tidak mengatur pengiriman tepat waktu" in reason_lower
        or "penjual gagal mengirimkan pesanan tepat waktu" in reason_lower):
        return "Seller Late Shipment"

    elif "lainnya" in reason_lower:
        return "Other"

    else:
        return "Other"


In [508]:
df_clean["cancellation_category"] = (
    df_clean["cancellation_reason"]
    .apply(categorize_cancellation_reason)
)

In [518]:
other_cancellation_reasons = (
    df_clean[
        (df_clean["order_status"] == "Batal")
        &
        (df_clean["cancellation_category"] == "Other")
    ]
    ["cancellation_reason"]
    .value_counts()
)

other_cancellation_reasons

,count
cancellation_reason,
Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,124
Dibatalkan oleh Pembeli. Alasan: Lainnya,2


In [509]:
df_clean[
    df_clean["order_status"] == "Batal"
]["cancellation_category"].value_counts()

,count
cancellation_category,
Other,126
Change Order,122
Change Address,107
Unpaid Order,77
Delivery Failed,61
Seller Responsiveness,5
Payment Difficulty,3
Voucher Issue,3
Seller Late Shipment,2


### Add the Cancellation Category to df_orders.

In [510]:
order_cancellation_category = (
    df_clean
    .groupby("order_id", as_index=False)
    .agg(
        cancellation_category=(
            "cancellation_category",
            "first"
        )
    )
)

In [511]:
df_orders = df_orders.merge(
    order_cancellation_category,
    on="order_id",
    how="left"
)

In [512]:
df_orders[
    df_orders["is_cancelled"] == 1
]["cancellation_category"].value_counts()

,count
cancellation_category,
Other,123
Change Order,121
Change Address,105
Unpaid Order,76
Delivery Failed,60
Seller Responsiveness,5
Payment Difficulty,3
Voucher Issue,3
Seller Late Shipment,2


In [513]:
df_orders[
    (df_orders["is_cancelled"] == 1)
    &
    (df_orders["cancellation_category"] == "Not Cancelled")
]

,order_id,order_status,order_created_at,order_date,payment_at,shipping_deadline,shipping_arranged_at,order_completed_at,payment_lead_time_minutes,processing_time_minutes,...,has_started_fulfillment,cancellation_stage,processing_category,is_late_shipment,is_returned,returned_items,total_items,unique_products,is_deadline_risk,cancellation_category


### Final Cancellation Summary

In [514]:
cancellation_summary = (
    df_orders[
        df_orders["is_cancelled"] == 1
    ]
    .groupby(
        [
            "cancellation_category",
            "cancellation_stage"
        ]
    )
    .agg(
        cancelled_orders=("order_id", "count")
    )
    .reset_index()
)

In [515]:
# adding pecentage
cancellation_summary["percentage"] = (
    cancellation_summary["cancelled_orders"]
    / cancelled_orders
    * 100
).round(2)

cancellation_summary = (
    cancellation_summary
    .sort_values(
        "cancelled_orders",
        ascending=False
    )
)

cancellation_summary

,cancellation_category,cancellation_stage,cancelled_orders,percentage
5,Other,Pre-Fulfillment,123,24.45
1,Change Order,Pre-Fulfillment,121,24.06
0,Change Address,Pre-Fulfillment,105,20.87
10,Unpaid Order,Pre-Fulfillment,76,15.11
3,Delivery Failed,Post-Fulfillment Initiation,60,11.93
9,Seller Responsiveness,Pre-Fulfillment,5,0.99
11,Voucher Issue,Pre-Fulfillment,3,0.60
7,Payment Difficulty,Pre-Fulfillment,3,0.60
2,Changed Mind,Pre-Fulfillment,2,0.40
6,Package Lost,Post-Fulfillment Initiation,2,0.40


### "Other" investigation

In [516]:
other_cancellation_reasons = (
    df_clean[
        (df_clean["order_status"] == "Batal")
        &
        (df_clean["cancellation_category"] == "Other")
    ]
    ["cancellation_reason"]
    .value_counts()
)

other_cancellation_reasons

,count
cancellation_reason,
Dibatalkan oleh Pembeli. Alasan: Alasan Lainnya,124
Dibatalkan oleh Pembeli. Alasan: Lainnya,2


## Final Business Insights

## Key Operational Insights

### 1. Order Completion Performance
A total of 3,153 unique orders were recorded during October 2025. Of these, 2,650 orders were completed and 503 were cancelled, resulting in an 84.05% completion rate and a 15.95% cancellation rate.

### 2. Fulfillment Processing Performance
Among 2,712 orders that entered fulfillment, the median processing time was approximately 18.88 hours, while 90% were processed within approximately 32.10 hours.

### 3. Critical Processing Cases Were Limited
26 processed orders, approximately 0.96%, were classified as Critical based on processing times above the 99th-percentile monitoring threshold.

### 4. Shipping Deadline Compliance Was High
Only 9 processed orders were arranged for shipping after their recorded marketplace deadline, resulting in a late-shipment rate of approximately 0.33% and an on-time shipping-arrangement rate of 99.67%.

### 5. Most Cancellations Occurred Before Fulfillment
Of 503 cancelled orders, 441 (87.67%) occurred before fulfillment began, while 62 (12.33%) occurred after fulfillment had already been initiated. Post-fulfillment cancellations may require additional operational attention because fulfillment or logistics activity has already started.

### 6. Return Incidence Was Relatively Limited
36 completed orders contained at least one returned item, corresponding to a return-order rate of approximately 1.36%. A total of 36 out of 3,513 ordered items were returned, representing an item return rate of approximately 1.02%.

### 7. Daily Order Volume Was Not Strongly Associated With Processing Time
The correlation between daily order volume and median processing time was approximately 0.076, indicating little linear association between workload volume and processing time during the observed period.

### Export Data

In [517]:
df_clean.to_csv(
    "vidrona_order_items_clean.csv",
    index=False
)

df_orders.to_csv(
    "vidrona_orders_clean.csv",
    index=False
)

daily_operations.to_csv(
    "vidrona_daily_operations.csv",
    index=False
)

cancellation_summary.to_csv(
    "vidrona_cancellation_summary.csv",
    index=False
)

reconciliation_summary.to_csv(
    "vidrona_reconciliation_summary.csv",
    index=False
)

operational_kpi.to_csv(
    "vidrona_operational_kpi.csv",
    index=False
)